packages/opentelemetry-core/src/default_span.ts

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Span,
+  SpanContext,
+  Status,
+  Attributes,
+  SpanKind,
+} from '@opentelemetry/types';
+
+/** This is a no-op implementation of Span. */
+export class DefaultSpan implements Span {
+  protected name = 'DefaultSpan'; // default
+  protected spanId = '';
+  protected parentSpanId = '';
+  protected kind = SpanKind.INTERNAL; // default
+  protected spanContext: SpanContext | null = null;
+
+  constructor() {
+    // TODO: https://github.com/open-telemetry/opentelemetry-js/pull/35
+    // this.spanId = randomSpanId();
+  }
+
+  context(): SpanContext {
+    return this._context();
+  }
+
+  setAttribute(key: string, value: unknown): this {
+    this._setAttributes({ key: value });
+    return this;
+  }
+
+  addEvent(name: string, attributes?: Attributes): this {
+    this._addEvent(name, attributes);
+    return this;
+  }
+
+  addLink(spanContext: SpanContext, attributes?: Attributes): this {
+    this._addLink(spanContext, attributes);
+    return this;
+  }
+
+  setStatus(status: Status): this {
+    this._setStatus(status);
+    return this;
+  }
+
+  updateName(name: string): this {
+    this._updateName(name);
+    return this;
+  }
+
+  end(endTime?: number): void {
+    this._end(endTime);
+  }
+
+  isRecordingEvents(): boolean {
+    return false;
+  }
+
+  // ---------------------------------------------------------------------- //
+  // Derived classes can choose to implement the below                      //
+  // ---------------------------------------------------------------------- //
+
+  // By default returns a no-op SpanContext.
+  protected _context(): SpanContext {
```

Comment:
What's the motivation for making these `protected` instead of just `private`? Do we expect people to subclass the no-op span?

Reply:
I was imaging impl/SDK (or even vender) would extend this `DefaultSpan` class and write their own implementation of these `protected` methods. May be we should consider to renaming this class to something like `BaseSpan`. WDYT?

Reply:
OK, I like the idea of calling it `BaseSpan`. Looking at it more closely, what's the motivation for the underscore-prefixed protected methods existing at all? Most of them just seem directly called from the public ones - why not just make them all be public?

Also, as a more minor thing, internally at Google we don't use underscores as prefixes for private methods, and neither does the TypeScript team itself (see https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#names). I guess protected methods are a bit different and people extending the class might not be using TypeScript so maybe the underscore makes sense.

In generally, I'm a bit wary of designing things to be inherited as I'd default to prefer [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance). That said, I think OpenTracing used a design similar to this so I'm open to it.

Reply:
> Looking at it more closely, what's the motivation for the underscore-prefixed protected methods existing at all (vs. just having subclases override the public methods)? Most of them just seem directly called from the public ones - why not just make them all be public?

Sure, removed the `protected` methods.

Reply:
Underscore prefixes are require in this case, for a few reasons:

* This is a library and not an application. This means most users will be using JavaScript and not TypeScript. An underscore is the only way they have to know this should not be called directly.
* The underscore variants allow a simpler inheritance model, since some behavior can be included in the public versions (non-prefixed). I agree this could be solved with composition instead.